### ISSUE TITLE
Add support for type parameters

### ISSUE THREAD

**Issue by dominikh** [dominikh/go-tools#1166]
Support Go 1.18's type parameters. Update all checks to handle new shapes of ASTs and new kinds of types. ~Haven't yet decided what to do with IR-based checks. Depending on various factors (time, features in go/types) we might have to skip analysis of packages that use type parameters for the time being (but not forever.)~

## Look for these patterns/stuff to do

- [x] Code that assumes that function calls with statically known functions must have a SelectorExpr or Ident as the called function. With type parameters, it can also be IndexExpr and IndexListExpr
- [x] Code that given a *types.Named uses Obj().Name() to get its unqualified name. For instantiated types this loses the instantiation. Use types.TypeString instead.
- [x] Code that assumes that IndexExpr can only be used for indexing collections
- [x] Make sure type switches don't work via exclusion. For example, existing code might assume that the first argument to `make` has to be a slice, map or channel, handle slices and maps one way, and assume that the only type left is channels. However, with type parameters, there might now be a fourth type, type parameters.
- [x] Audit all uses of code.IsOfType and typeutil.IsType
- [x] Method receiver types are no longer just StarExpr and Ident, they may also be IndexExpr and IndexListExpr; and StarExpr no longer just wraps Idents.
- [x] Audit every type assertion on the result of `Underlying`. ~Everything might be an interface now.
- [x] IR-based analyzes that print function names shouldn't print the names of generic wrappers. The user shouldn't see `f$generic#0 is a pure function but its return value is ignored`
- [x] Fix all our test cases so they compile on Go 1.17

## Vet checks, handle new syntax and types, write new test cases

- [x] SA1000
- [x] SA1001
- [x] SA1002
- [x] SA1003
- [x] SA1004
- [x] SA1005
- [x] SA1006
- [x] SA1007
- [x] SA1008
- [x] SA1010
- [x] SA1011
- [x] SA1012
- [x] SA1013
- [x] SA1014
- [x] SA1015
- [x] SA1016
- [x] SA1017
- [x] SA1018
- [x] SA1019
- [x] SA1020
- [x] SA1021
- [x] SA1023
- [x] SA1024
- [x] SA1025
- [x] SA1026
- [x] SA1027
- [x] SA1028
- [x] SA1029
- [x] SA1030
- [x] SA2000
- [x] SA2001
- [x] SA2002
- [x] SA2003
- [x] SA3000
- [x] SA3001
- [x] SA4000
- [x] SA4001
- [x] SA4003
- [x] SA4004
- [x] SA4005
- [x] SA4006
- [x] SA4008
- [x] SA4009
- [x] SA4010
- [x] SA4011
- [x] SA4012
- [x] SA4013
- [x] SA4014
- [x] SA4015
- [x] SA4016
- [x] SA4017
- [x] SA4018
- [x] SA4019
- [x] SA4020
- [x] SA4021
- [x] SA4022
- [x] SA4023
- [x] SA4024
- [x] SA4025
- [x] SA4026
- [x] SA4027
- [x] SA4028
- [x] SA4029
- [x] SA4030
- [x] SA4031
- [x] SA5000
- [x] SA5001
- [x] SA5002
- [x] SA5003
- [x] SA5004
- [x] SA5005
- [x] SA5007
- [x] SA5008
- [x] SA5009
- [x] SA5010
- [x] SA5011
- [x] SA5012
- [x] SA6000
- [x] SA6001
- [x] SA6002
- [x] SA6003
- [x] SA6005
- [x] SA9001
- [x] SA9002
- [x] SA9003
- [x] SA9004
- [x] SA9005
- [x] SA9006
- [x] SA9007
- [x] SA9008
- [x] S1000
- [x] S1001
- [x] S1002
- [x] S1003
- [x] S1004
- [x] S1005
- [x] S1006
- [x] S1007
- [x] S1008
- [x] S1009
- [x] S1010
- [x] S1011
- [x] S1012
- [x] S1016
- [x] S1017
- [x] S1018
- [x] S1019
- [x] S1020
- [x] S1021
- [x] S1023
- [x] S1024
- [x] S1025
- [x] S1028
- [x] S1029
- [x] S1030
- [x] S1031
- [x] S1032
- [x] S1033
- [x] S1034
- [x] S1035
- [x] S1036
- [x] S1037
- [x] S1038
- [x] S1039
- [x] S1040
- [x] ST1000
- [x] ST1001
- [x] ST1003
- [x] ST1005
- [x] ST1006
- [x] ST1008
- [x] ST1011
- [x] ST1012
- [x] ST1013
- [x] ST1015
- [x] ST1016
- [x] ST1017
- [x] ST1018
- [x] ST1019
- [x] ST1020
- [x] ST1021
- [x] ST1022
- [x] ST1023
- [x] QF1001
- [x] QF1002
- [x] QF1003
- [x] QF1004
- [x] QF1005
- [x] QF1006
- [x] QF1007
- [x] QF1008
- [x] QF1009
- [x] QF1010
- [x] QF1011
- [x] QF1012
- [x] U1000 - tentatively marked done
  - [x] Don't flag unused methods or fields on instantiated types.

**Comment by ainar-g** [comment]
I'm not sure if it's too early to comment on the state of the `generics` branch, so apologies if it's nowhere near ready yet, but I currently have the following crash:

```sh
staticcheck ./... github.com/AdguardTeam/AdGuardDNS/internal/dnsserver/...
```

```none
panic: unreachable: *types.Union

goroutine 754 [running]:
honnef.co/go/tools/unused.(*graph).typ(0xc006099100, {0x9d3378?, 0xc001b61b90?}, {0x0, 0x0?})
        /home/ainar/go/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1523 +0x8d9
honnef.co/go/tools/unused.(*graph).typ(0xc006099100, {0x9d31e8?, 0xc000bfde60?}, {0x9d3350, 0xc001ea45d0?})
        /home/ainar/go/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1490 +0xd7b
honnef.co/go/tools/unused.(*graph).typ(0xc006099100, {0x9d3350?, 0xc001ea45d0?}, {0x0, 0x0?})
        /home/ainar/go/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1521 +0x495
[…]
```

**Comment by dominikh** [comment]
At this point it's fine to report any crashes, and bugs in `unused`. Thanks.

**Comment by 89z** [comment]
https://github.com/dominikh/go-tools/issues/1201

**Comment by 89z** [comment]
https://github.com/dominikh/go-tools/issues/1203

**Comment by 89z** [comment]
https://github.com/dominikh/go-tools/issues/1200

**Comment by vngzs** [comment]
Similar error here:

```shell
$ cat main.go
package main

import (
        "fmt"

        "golang.org/x/exp/maps"
)

func main() {
        set := make(map[string] bool)
        set["hello"] = true
        set["world"] = true
        fmt.Println(maps.Keys(set))
}
```

```shell
$ staticcheck main.go
panic: unreachable: *types.Union

goroutine 128 [running]:
honnef.co/go/tools/unused.(*graph).typ(0xc00068e740, {0x9cf1d8?, 0xc0005824b0?}, {0x0, 0x0?})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1523 +0x8d9
honnef.co/go/tools/unused.(*graph).typ(0xc00068e740, {0x9cf048?, 0xc0001ba300?}, {0x9cf1b0, 0xc000095e90?})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1490 +0xd7b
honnef.co/go/tools/unused.(*graph).typ(0xc00068e740, {0x9cf1b0?, 0xc000095e90?}, {0x0, 0x0?})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1521 +0x495
honnef.co/go/tools/unused.(*graph).instructions(0xc00068e740, 0xc000617400)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1614 +0x285
honnef.co/go/tools/unused.(*graph).function(0xc00068e740, 0xc000617400)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1334 +0xbe
honnef.co/go/tools/unused.(*graph).instructions.func1({0x9d3078?, 0xc000617400?})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1594 +0x125
honnef.co/go/tools/unused.walkPhi({0x9d3078?, 0xc000617400?}, 0x0?)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1815 +0x11f
honnef.co/go/tools/unused.(*graph).instructions(0xc00068e740, 0xc0006172c0)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1584 +0x1168
honnef.co/go/tools/unused.(*graph).function(0xc00068e740, 0xc0006172c0)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1334 +0xbe
honnef.co/go/tools/unused.(*graph).entry(0xc00068e740, 0xc000140280)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:1168 +0x18fa
honnef.co/go/tools/unused.run(0xc000165d40)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/unused/unused.go:536 +0x359
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000619f50, {0x9d19d8?, 0xc000131cc0})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9d19d8?, 0xc000131cc0}, {0x9d19d8?, 0xc0001ce3c0?}, 0x0?, 0x0, 0xc0005f8e40)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:857 +0x102
honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers(0xc0001ca1e0, 0xc00017e480, 0xc0001cbd60)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:1105 +0x766
honnef.co/go/tools/lintcmd/runner.(*subrunner).doUncached(0xc0001ca1e0, 0xc00017e480)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:728 +0x1c9
honnef.co/go/tools/lintcmd/runner.(*subrunner).do(0xc0001ca1e0, {0x9d1a20?, 0xc00017e480})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:584 +0x78d
honnef.co/go/tools/lintcmd/runner.(*Runner).Run.func2({0x9d1a20?, 0xc00017e480?})
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:1308 +0x30
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9d1a20?, 0xc00017e480}, {0x9d1a20?, 0xc00017e360?}, 0xc00005afd0?, 0xc000393c60, 0xc00037c030)
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*Runner).Run
        ~/pkg/mod/honnef.co/go/tools@v0.3.0-0.dev.0.20220309174655-774e2c12739b/lintcmd/runner/runner.go:1307 +0x525
```

**Comment by dominikh** [comment]
@ainar-g @vngzs that particular crash should be fixed now.

**Comment by ainar-g** [comment]
@dominikh, confirming, that one is fixed, and no new ones so far.

**Comment by Al2Klimov** [comment]
@dominikh Do you consider https://github.com/dominikh/go-tools/tree/generics already usable? (I.e. I'd check out – say – the head commit in GH actions.)

**Comment by mvdan** [comment]
It's ready for testing or early users, but not ready for general use given that the release isn't out yet.

**Comment by dominikh** [comment]
Closed by 23e1086..a34f635. The release will follow soon.
